include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+ * of those objects to the application. It is acceptable to call this function
+ * to add another linked buffer to the send queue, even if the "send done"
+ * callback for a previous invocation of this function has not yet fired.
+ *
+ * Note that @p aBuffer should not be chained; its @p mNext field should be
+ * @p NULL. If additional data will be added right after this call, then the
+ * @p OT_TCP_ACTIVE_SEND_MORE_TO_COME flag should be used as a hint to the TCP
+ * implementation.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure representing the TCP endpoint on which to send data.
+ * @param[in]  aBuffer    A pointer to the linked buffer chain referencing data to add to the send buffer.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully added data to the send buffer.
+ * @retval OT_ERROR_FAILED  Failed to add data to the send buffer.
+ */
+otError otTcpActiveSendByReference(otInstance *aInstance, otTcpActiveSocket *aSocket, otLinkedBuffer *aBuffer, uint32_t aFlags);
+
+/**
+ * Adds data to the send buffer by extending the length of the final
+ * otLinkedBuffer in the send buffer by the specified amount. If the send
+ * buffer is empty, then the operation fails.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure representing the TCP endpoint on which to send data.
+ * @param[in]  aNumBytes  The number of bytes by which to extend the length of the final linked buffer.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully added data to the send buffer.
+ * @retval OT_ERROR_FAILED  Failed to add data to the send buffer.
+ */
+otError otTcpActiveSendByExtension(otInstance *aInstance, otTcpActiveSocket *aSocket, size_t aNumBytes, uint32_t aFlags);
+
+/**
+ * Provides the application with a linked buffer chain referencing data
+ * currently in the TCP receive buffer. The linked buffer chain is valid until
+ * the "receive ready" callback is next invoked, or until the next call to
+ * @p otTcpActiveReceiveContiguify or @p otTcpActiveCommitReceive.
+ *
+ * @param[in]   aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]   aSocket    A pointer to the active TCP socket structure representing the TCP endpoint on which to receive data.
+ * @param[out]  aBuffer    A pointer to the linked buffer chain referencing data currently in the receive buffer.
+ *
+ * @retval OT_ERROR_NONE    Successfully completed the operation.
+ * @retval OT_ERROR_FAILED  Failed to complete the operation.
+ */
+otError otTcpActiveReceiveByReference(otInstance *aInstance, otTcpActiveSocket *aSocket, const otLinkedBuffer **aBuffer);
```

Comment:
Who owns this `otLinkedBuffer` structure?

Can user pass a pointer to a local `otLinkedBuffer` to receive the data?

```suggestion
otError otTcpActiveReceiveByReference(otInstance *aInstance, otTcpActiveSocket *aSocket, otLinkedBuffer *aBuffer);
```

Reply:
The system owns the `otLinkedBuffer` and returns a pointer to it. `aBuffer` is a double pointer because the function sets it to a pointer to an `otLinkedBuffer` that it owns. That's why it's marked with `@param[out]` in the comment above the function.

Reply:
I guess the `otLinkedBuffer` is allocated from the `TCP receive buffer`.

Can ALP locate a `otLinkedBuffer` itself and receive data to it (TCP set it to reference the data in `receive buffer` ) ?

```c++
otLinkedBuffer recvBuffer;

otTcpActiveReceiveByReference(aInstance, aSocket, &recvBuffer);

// Now data is referenced by recvBuffer.mData and recvBuffer.mLength
```

I am not sure if this can free TCP from having to allocating `otLinkedBuffer` from `receive buffer`. Thoughts?

Reply:
I considered that approach, but it's hard for the application to know how many `otLinkedBuffer`s to use.

But I think having the system allocate `otLinkedBuffer`s is not very hard. If we implement the receive queue as a queue of `otMessage`s, then we follow your suggestion of including an `otLinkedBuffer` in `otMessageBuffer`. Then this problem goes away.

If we use the approach I suggested, which is to use a flat circular buffer for the receive buffer (in order to avoid `otMessage` livelock) then the system needs at most two `otLinkedBuffer`s to reference the data in the circular buffer. These two `otLinkedBuffer`s can just be fields in `otTcpActiveSocket`.

So in summary I don't think it's difficult for the TCP implementation to allocate these `otLinkedBuffer`s.

Reply:
> I considered that approach, but it's hard for the application to know how many otLinkedBuffers to use.

Note that ALP can allocate the `otLinkedBuffer` locally in stack.

> If we use the approach I suggested, which is to use a flat circular buffer for the receive buffer (in order to avoid `otMessage` livelock) then the system needs at most two `otLinkedBuffer`s to reference the data in the circular buffer. These two `otLinkedBuffer`s can just be fields in `otTcpActiveSocket`.

Do we keep OOO segments in the receive buffer? If so, we will need one `otLinkedBuffer` for in-order data and one `otLinkedBuffer` for each OOO segment? I am not very clear how we are going to manage OOO segments in `receive buffer`.